### ISSUE TITLE
Map.getLoadingOrNotReady() should check source.getState() === 'loading'

### ISSUE THREAD

**Issue by jcphill** [openlayers/openlayers#16483]
`Map.getLoadingOrNotReady()` is used to generate `loadstart` and `loadend` events. It checks if any sources have their loading property true but not all sources set this value (this is noted in the comments). In particular `GeoTIFFSource` does not modify its loading property but it does update its state, so checking `source.getState() === 'loading'` in addition to `source.loading` should produce the correct behavior. Despite the function name, I would not check `source.getState() !== 'ready'` as I assume `loadend` should be issued even if loading stops due to an error.

https://github.com/openlayers/openlayers/blob/c57f740aa97fdb56fcae616030a49f415b44cfd6/src/ol/Map.js#L960-L980

**Comment by jcphill** [comment]
`Map.getLoadingOrNotReady()` should also call `state.layer.isVisible()` rather than just checking `state.visible`.

**Comment by jcphill** [comment]
This is the comment that mentions that not all sources set loading:
https://github.com/openlayers/openlayers/blob/d58005cc2ab3777d2225eaa211ced4be82d595b4/src/ol/source/Source.js#L76-L81

**Comment by ahocevar** [comment]
That's two different things. The source's `state_` keeps track of whether the source is `"ready"` to be used, i.e. when it is fully configured. The source's `loading` property indicates whether the source is loading data.

@jcphill What was the problem you were running into that made you create this issue?

**Comment by jcphill** [comment]
I have a loading spinner copied from https://openlayers.org/en/latest/examples/load-events.html on a WebGLTile layer with a GeoTIFF source that reads a 50 GB cloud optimized GeoTIFF file and it takes several seconds to read the image directories before the source becomes "ready". Without these changes the spinner stops after the base layer is displayed, only to restart several seconds layer when the GeoTIFF overlay starts rendering.

The doc string below suggest that `getLoadingOrNotReady()` should return true when the GeoTIFF source is reading header data even if it is not ready to handle requests from the layer.

https://github.com/openlayers/openlayers/blob/d58005cc2ab3777d2225eaa211ced4be82d595b4/src/ol/Map.js#L960-L963

A secondary complication is that the WebGLTile layer is only visible at high zoom values (minZoom: 14), so the spinner should not display while the source loads image directories if the map zoom level is below minZoom, which is one of the things that `state.layer.isVisible()` checks.

**Comment by ahocevar** [comment]
Thanks for the explanation, @jcphill. Agreed on both. Would you be able to create a pull request, with a fix and with tests?

**Comment by jcphill** [comment]
I can create a pull request but I don't see any existing tests for `getLoadingOrNotReady()` and I have no idea how I would create one given that this is a transient state.

I also don't like how `state.layer.isVisible()` itself searches the same layerStatesArray and would want to refactor `isVisible()` to take layerState as an optional argument to avoid O(n^2) operations but I don't understand most of the cases that function is trying to handle.
https://github.com/openlayers/openlayers/blob/d58005cc2ab3777d2225eaa211ced4be82d595b4/src/ol/layer/Layer.js#L315-L322

**Comment by jcphill** [comment]
Going further down the rabbit hole, if the map view is initialized from `GeoTIFFSource.getView()`, which returns a promise, then `map.frameState_` is null and loadstart isn't dispatched until the source is ready anyway.
https://github.com/openlayers/openlayers/blob/d58005cc2ab3777d2225eaa211ced4be82d595b4/src/ol/Map.js#L1209-L1229

**Comment by ahocevar** [comment]
Thinking more about this, I think as a first step it would be better if you could create a minimal example that shows the problems. The visibility problem could be solved without calling `isVisible()` on the layer, by iterating layer groups on the map instead. Either way, it would be a bigger change, and I'd like to have an example that shows the problems before we start drafting a solution.

**Comment by jcphill** [comment]
Try this: https://deploy-preview-16535--ol-site.netlify.app/en/latest/examples/cog-load-events.html

The observed event order "cogready loadstart loadend" would ideally be "loadstart cogready loadend".

**Comment by ahocevar** [comment]
> The observed event order "cogready loadstart loadend" would ideally be "loadstart cogready loadend".

The order is exactly what I would expect. See my [comment](https://github.com/openlayers/openlayers/pull/16535#issuecomment-2576373196).

**Comment by jcphill** [comment]
Yes, the code as currently implemented does what you as the developer expect, but from a user perspective downloading GeoTIFF image directories before the source is fully configured is not materially different from downloading and rendering image tiles after, so it would be nice if loadstart was fired to activate the spinner while waiting for the source to be configured.

**Comment by ahocevar** [comment]
Not sure if this is what everyone would expect. The map's `loadstart` and `loadend` events are meant to indicate that data for rendering the map is loading. In order not to mix that with waiting for sources to be `ready`, maybe a new `ready` event on the map would be a better solution?

**Comment by ahocevar** [comment]
> Is there a use case for the loadstart and loadend events that excludes pre-ready source loading?

Sort of. A source might not have data for the current viewport's extent or resolution, and we do not know that before the source is ready. We don't know if data from that source will be requested at all for the current viewport.

**Comment by jcphill** [comment]
But what is the use case for that distinction? I would prefer treating pending (meta)data as relevant to the viewport unless there is evidence that it is not (e.g., minZoom/maxZoom on the layer). A source that does have data for the current viewport might render entirely as transparent so you know nothing for sure until you are done rendering.

**Comment by ahocevar** [comment]
@jcphill Following up on

> Map.getLoadingOrNotReady() should also call state.layer.isVisible() rather than just checking state.visible.

I looked at the code, and the `LayerGroup#getLayerStatesArray` takes min/maxZoom and group visibility into account already. The `layer.isVisible()` method's purpose is to get the same when the context of the whole frameState's layerStatesArray is not available, i.e. when an application needs that information.

What makes you think otherwise?

If I'm correct, #16530 should already be the complete fix.

**Comment by jcphill** [comment]
With #16530 applied the spinner is shown even if the map zoom is outside the min/maxZoom of the layer when `source.getState() === 'loading'`. I assume "the context of the whole frameState's layerStatesArray is not available" until the layer source is ready. Likely `layer.isVisible()` or similar only needs to be checked when `source.getState() === 'loading'`, something like:
`if (source && (source.loading || (source.getState() === 'loading' && state.layer.isVisible())))`

**Comment by ahocevar** [comment]
Probably
```js
if (source && (source.loading || (source.getState() === 'loading' && state.visible)))
```
Can you update the pull request, @jcphill ?

**Comment by jcphill** [comment]
That would change nothing since `state.visible` is checked already at https://github.com/openlayers/openlayers/blob/c57f740aa97fdb56fcae616030a49f415b44cfd6/src/ol/Map.js#L967
and the problem is that `state.visible` is not reliable before the source is ready.

**Comment by ahocevar** [comment]
You're looking at a code path that is not being used here. This is the relevant code section:

https://github.com/openlayers/openlayers/blob/main/src/ol/layer/Group.js#L305-L348

If you have tests, then share them. But I think we'll be back at a previous cycle of this discussion, where I said you cannot know if you'll be loading data from a source before you know which zoom levels and extent the source provides data for.

**Comment by jcphill** [comment]
That group implementation of `getLayerStatesArray` recursively calls `getLayerStatesArray` on its member layers:
https://github.com/openlayers/openlayers/blob/b0c7d63cc551b7ea6fe1ef402c6f8d5e24abc2f4/src/ol/layer/Group.js#L309-L311
which for non-group layers calls `getLayerState`:
https://github.com/openlayers/openlayers/blob/b0c7d63cc551b7ea6fe1ef402c6f8d5e24abc2f4/src/ol/layer/Layer.js#L197-L201
which calls `getVisible':
https://github.com/openlayers/openlayers/blob/b0c7d63cc551b7ea6fe1ef402c6f8d5e24abc2f4/src/ol/layer/Base.js#L168
and the only change is to adjust for group visibility:
https://github.com/openlayers/openlayers/blob/b0c7d63cc551b7ea6fe1ef402c6f8d5e24abc2f4/src/ol/layer/Group.js#L321

At least for a `WebGLTileLayer` if I set min/maxZoom in the constructor then I can prove that the layer is not visible at zoom levels outside of that range without loading any source data.

**Comment by ahocevar** [comment]
Same for minZoom and maxZoom on the layer state:

https://github.com/openlayers/openlayers/blob/b0c7d63cc551b7ea6fe1ef402c6f8d5e24abc2f4/src/ol/layer/Group.js#L330-L331

Nice. These could (and should) be used here, in addition to `state.visible`, but no need to call `state.layer.isVisible()`.
